### ISSUE TITLE
Compress files parallel to the main thread

### ISSUE THREAD

**Issue by bibusdax** [lochmueller/staticfilecache#164]
## Feature Request

**Is your feature request related to a problem? Please describe.**
Currently with all compression options enabled, the rendering of an uncached page is very slow.
This is because the compressions are done in sequence and not in parallel.
In my case I have to rebuild more than 30'000 pages when I clear tx_staticfilecache.
Because of the compression being sequential, this takes several days and if the compression would be done in parallel, I could have a significant speed improvement in building the cache.
Also visitors to the page, when they open an uncached page currently have to wait more than 20 seconds, which is way too slow. So putting the compression into a background task, would help here, too.

**Describe the solution you'd like**
I recommend that the compression "Generators" would be rewritten in a way that the compression part itself is either "Threaded" or a shell script is being called in the background to compress the index.html file.
Kown problem: pthreads does not work in the web server environment. So the necessity to enforce page rendering via PHP called from the CLI. eventually.

**Describe alternatives you've considered**
There are two ideas:
a) Turn off all compression to gain speed in rendering and have some Shell-Scripts do the compression.
b) Rewrite the code for the "Generators" of the compressed files.

**Teachability, Documentation, Adoption, Migration Strategy**
If paralellization can be done in the web server environment, no big adaption would be necessary.
However if that can only be done via Shell/Scheduler command, then instructions are necessary to create such a task.

**Comment by bibusdax** [comment]
I tested parallelization and compression is so fast that this is not the bottleneck I am looking for.
So I close this issue.

**Comment by lochmueller** [comment]
Hey @bibusdax 
compression do not need much time. I suggest to run some blackfire.io profiles to get information why your website is slow. StaticFileCache itself has a very small footprint in the request time.
Regards,
Tim

**Comment by bibusdax** [comment]
Dear Tim,

Thank you very much for your quick reply.
I checked the timings of the cache and you are right, that is not the guilty part slowing everything down.

The guilty part is most probably our SOLR extension.
There I will have to check with DKD.

I am just in general cleanup mode of our TYPO3 CMS. We upgraded to 9LTS recently and now it is cleanup time.

You are very helpful and generous.

Thank you very much !

Best regards,

Patrick Daxboeck
IT-Consultant Web
Certified Web Project Manager

BIBUS HOLDING AG
Allmendstrasse 26
8320 Fehraltorf
Switzerland
Direct   +41 44 877 50 34
Mobile  +41 76 331 00 31
E-Mail   dax@bibus.ch
Internet www.bibus-holding.ch



Von: Tim [mailto:notifications@github.com]
Gesendet: Mittwoch, 5. Juni 2019 14:17
An: lochmueller/staticfilecache
Cc: Daxboeck Patrick; Mention
Betreff: Re: [lochmueller/staticfilecache] Compress files parallel to the main thread (#164)


Hey @bibusdax<https://github.com/bibusdax>
compression do not need much time. I suggest to run some blackfire.io profiles to get information why your website is slow. StaticFileCache itself has a very small footprint in the request time.
Regards,
Tim
